src/device/rpi3/timer.go

```diff
@@ -0,0 +1,335 @@
+
+func BComTimerInit(interval uint32) {
+	bcomCurVal = volatile.LoadUint32((*uint32)(TIMER_CLO))
+	bcomCurVal += interval
+	volatile.StoreUint32((*uint32)(TIMER_C1), bcomCurVal)
+}
+
+func BComHandleTimerIRQ(interval uint32) {
+	bcomCurVal += interval
+	volatile.StoreUint32((*uint32)(TIMER_C1), bcomCurVal)
+	volatile.StoreUint32((*uint32)(TIMER_CS), TIMER_CS_M1)
+	print("IRQ handler: ")
+	UART0Hex(bcomCurVal)
+}
+
+func EnableTimerIRQ() {
+	Asm("msr daifclr, #2")
+}
+
+func DisableTimerIRQ() {
+	Asm("msr daifset, #2")
+}
+
+func EnableInterruptController() {
+	volatile.StoreUint32((*uint32)(ENABLE_IRQS_1), SYSTEM_TIMER_IRQ_1)
+}
+
+// the exported name is what connects it to the interrupt vector
+//go:export unexpectedInterrupt
+func UnexpectedInterrupt(n int, esr uint64, address uint64) {
+	if n < 0 || n >= len(unexpectedInterruptNames) {
+		print("unexpected interrupt called with bad index", n, "\n")
+		return
+	}
+	print("unexpected interrupt:", unexpectedInterruptNames[n], " esr:", esr, " address 0x")
+	UART0Hex64(address)
+	sp := ReadRegister("sp")
+	print("sp is ")
+	UART0Hex64(uint64(sp))
+	Asm("eret")
+}
+
+// get the counter frequency
+func CounterFreq() uint32 {
+	var val uint32
+	AsmFull(`mrs x28, cntfrq_el0
+		str x28,{val}`, map[string]interface{}{"val": &val})
+	return val
+}
+
+//returns 0 when it doesn't want the timer anymore
+var userCallbackFunc func(uint32, uint64) uint32
+
+// clears the interrupt and writes target value (which is a time distance from now)
+func SetCounterTargetInterval(val uint32, fn func(uint32, uint64) uint32) {
+	AsmFull(`mov x28,{val}
+		msr cntv_tval_el0, x28`, map[string]interface{}{"val": val})
+	userCallbackFunc = fn
+}
+
+func CounterTargetVal() uint32 {
+	var val uint32
+	AsmFull(`mrs x28,cntv_tval_el0
+		str x28,{val}`, map[string]interface{}{"val": &val})
+	return val
+}
+
+func Core0CounterToCore0Irq() {
+	volatile.StoreUint32((*uint32)(CORE0_TIMER_IRQCNTL), 0x08)
+}
+
+func EnableCounter() {
+	Asm(`orr x27, xzr, #1
+		msr cntv_ctl_el0, x27`)
+}
+func UDisableCounter() {
+	Asm("msr cntv_ctl_el0, #0")
+}
+
+func WaitForInterrupt() {
+	Asm("wfi")
+}
+
+func Core0TimerPending() uint32 {
+	return volatile.LoadUint32((*uint32)(CORE0_IRQ_SOURCE))
+}
+
+func VirtualTimer() uint64 {
+	var val uint64
+	AsmFull(`mrs x28, cntvct_el0
+		str x28,{val}`, map[string]interface{}{"val": &val})
+	return val
+}
+
+// the export name is what connects it to the list of interrupt handlers
+//go:export irq_el1h_handler
+//go:extern
+func HandleTimerInterrupt() {
+	DisableTimerIRQ()
+	//defer EnableTimerIRQ() --->  crashes the compiler
+	//defer func() {
+	// EnableTimerIRQ()
+	//}()                    ---> also crashes the compiler
+	// error:
+	// tinygo build --target=rpi3_bl -o kernel8.elf .
+	// Assertion failed: (From->getType() == To->getType()), function replaceDominatedUsesWith, file /Users/iansmith/tinygo.src/src/github.com/tinygo/tinygo/llvm-project/llvm/lib/Transforms/Utils/Local.cpp, line 2401.
```

Comment:
This is most likely fixed now, see #553.

Reply:
Fixed.